gateway/src/apicast/policy/ip_check/ip_check.lua

```diff
@@ -0,0 +1,75 @@
+local iputils = require("resty.iputils")
+
+local policy = require('apicast.policy')
+local _M = policy.new('IP check policy')
+
+local new = _M.new
+
+local function parse_cidrs(cidrs)
+  return iputils.parse_cidrs(cidrs or {})
+end
+
+local function request_ip_addr()
+  return ngx.var.remote_addr
+end
+
+local function req_ip_in_range(cidrs)
+  return iputils.ip_in_cidrs(request_ip_addr(), cidrs)
+end
+
+local function deny_request()
+  ngx.status = ngx.HTTP_FORBIDDEN
+  ngx.say('IP address not allowed')
+  ngx.exit(ngx.status)
+end
+
+local function apply_whitelist(self)
+  if not req_ip_in_range(self.whitelisted_ips) then
+    deny_request()
+  end
+end
+
+local function apply_blacklist(self)
+  if req_ip_in_range(self.blacklisted_ips) then
+    deny_request()
+  end
+end
+
+local function log_mutually_exclusive_err()
+  ngx.log(ngx.ERR, 'Invalid config. Blacklist and whitelist are mutually exclusive')
+end
+
+local noop = function() end
+
+local function check_ips_function(blacklisted_ips, whitelisted_ips)
+  local blacklist = #blacklisted_ips > 0
+  local whitelist = #whitelisted_ips > 0
+
+  if blacklist and whitelist then
+    log_mutually_exclusive_err()
```

Comment:
I didn't think about the use case that you mentioned.

I was thinking about cases where the 2 sets of IPs do not overlap. For example, blacklist '1.2.3.0/28' and whitelist '2.3.4.0/28'. What should we do with an IP like '3.3.3.3' ? It's not blacklisted, but not whitelisted either. So we'd need to complicate things a bit to define what happens in scenarios like that one.

Reply:
It's not clear to me why whitelisting or blacklisting should take precedence over the other.

I think the best option is to add a "type" attr in the policy config, as you suggested. That would avoid the "precedence" problem. Also, it would allow whitelisting a group of IPs and blacklisting one in that group by chaining 2 instances of the policy.

Reply:
Agreed 👍

Reply:
Done :+1: